nhs/data/join_on_sa1.py

```diff
@@ -0,0 +1,219 @@
+                    .otherwise(pl.col("x"))
+                    .alias("x"),
+                    pl.when(pl.col("area_code") == area_code)
+                    .then(pl.Series(shuffled_pairs["y"]))
+                    .otherwise(pl.col("y"))
+                    .alias("y"),
+                ]
+            )
+
+    return model_df
+
+
+def visualize_results(final_result_df: pl.DataFrame, area_codes: List[str]) -> None:
+    """Visualize the original and final positions of points."""
+    for area_code in area_codes:
+        area_result_df = final_result_df.filter(pl.col("area_code") == area_code)
+
+        plt.figure(figsize=(12, 6))
+        plt.subplot(1, 2, 1)
+        plt.scatter(
+            area_result_df["original_lon"],
+            area_result_df["original_lat"],
+            c=area_result_df["attribute"].to_numpy(),
+            cmap="tab10",
+            s=50,
+            alpha=0.7,
+        )
+        plt.title(f"Original Positions for Area Code: {area_code}")
+        plt.xlabel("Original Longitude")
+        plt.ylabel("Original Latitude")
+
+        plt.subplot(1, 2, 2)
+        plt.scatter(
+            area_result_df["x"],
+            area_result_df["y"],
+            c=area_result_df["attribute"].to_numpy(),
+            cmap="tab10",
+            s=50,
+            alpha=0.7,
+        )
+        plt.title(
+            f"Final Positions after Random Distribution for Area Code: {area_code}"
+        )
+        plt.xlabel("Final Longitude")
+        plt.ylabel("Final Latitude")
+
+        plt.tight_layout()
+        plt.show()
+
+
+def main() -> None:
+    config_path = "configurations.yaml"
+    config = read_config(config_path)
+
+    shapefile_path = config["shapefile_path"]
+    csv_path = config["csv_path"]
+    crs = config["crs"]
+
+    total_rows = read_shapefile(shapefile_path)
+    houses_df = read_csv(csv_path)
+
+    pnts_gdf = prepare_points(houses_df, crs)
+    map_data = gpd.read_file(shapefile_path).to_crs(pnts_gdf.crs)
+
+    num_cores = config.get("num_cores", 4)  # Default to 4 if not specified
+    final_result = parallel_process(pnts_gdf, map_data, num_cores)
+
+    if (
+        final_result["area"].is_null().any()
+        or final_result["area_code"].is_null().any()
+    ):
+        logger.warning(
+            "Some points were not assigned to an SA1 area even after nearest join. Please check your data."
+        )
+
+    houses_with_areas = houses_df.join(final_result, on=["x", "y"], how="left")
+
+    output_csv_path = config["output_csv_path"]
+    houses_with_areas.write_csv(output_csv_path)
+    logger.info(f"CSV file saved to: {output_csv_path}")
+
+    missing_area_count = final_result["area"].is_null().sum()
+    logger.info(f"Number of missing values in the 'area' column: {missing_area_count}")
+
+    houses_with_attributes = houses_with_areas.with_columns(
+        pl.Series("attribute", np.random.randint(1, 11, size=len(houses_with_areas)))
+    )
```

Comment:
What does this logic represent? 

I thought houses_with_attributes would be the specific types of building from the GNAF as described in the config.

It looks like it is not used or referenced later in the code the attribute column.

Reply:
Was using that in visualization testing to see if the Schelling model was working, those attributes were just random placeholders. I've removed them now.